Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.h

```diff
@@ -0,0 +1,73 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if ENABLE(SERVICE_WORKER)
+
+#include "BackgroundFetch.h"
+
+namespace WebCore {
+
+class BackgroundFetchStore;
+class ResourceResponse;
+class SWServer;
+
+class BackgroundFetchEngine : public CanMakeWeakPtr<BackgroundFetchEngine> {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    explicit BackgroundFetchEngine(SWServer&);
+
+    using ExceptionOrBackgroundFetchInformationCallback = CompletionHandler<void(Expected<BackgroundFetchInformation, ExceptionData>&&)>;
+    void startBackgroundFetch(SWServerRegistration&, const String&, Vector<BackgroundFetchRequest>&&, BackgroundFetchOptions&&, ExceptionOrBackgroundFetchInformationCallback&&);
+    void backgroundFetchInformation(SWServerRegistration&, const String&, ExceptionOrBackgroundFetchInformationCallback&&);
+    using BackgroundFetchIdentifiersCallback = CompletionHandler<void(Vector<String>&&)>;
+    void backgroundFetchIdentifiers(SWServerRegistration&, BackgroundFetchIdentifiersCallback&&);
+    using AbortBackgroundFetchCallback = CompletionHandler<void(bool)>;
+    void abortBackgroundFetch(SWServerRegistration&, const String&, AbortBackgroundFetchCallback&&);
+    using MatchBackgroundFetchCallback = CompletionHandler<void(Vector<BackgroundFetchRecordInformation>&&)>;
+    void matchBackgroundFetch(SWServerRegistration&, const String&, RetrieveRecordsOptions&&, MatchBackgroundFetchCallback&&);
+    using RetrieveRecordResponseCallback = BackgroundFetch::RetrieveRecordResponseCallback;
+    void retrieveRecordResponse(BackgroundFetchRecordIdentifier, RetrieveRecordResponseCallback&&);
+    using RetrieveRecordResponseBodyCallback = BackgroundFetch::RetrieveRecordResponseBodyCallback;
+    void retrieveRecordResponseBody(BackgroundFetchRecordIdentifier, RetrieveRecordResponseBodyCallback&&);
+
+    void remove(SWServerRegistration&);
+
+private:
+    void notifyBackgroundFetchUpdate(BackgroundFetchInformation&&);
+
+    WeakPtr<SWServer> m_server;
+    Ref<BackgroundFetchStore> m_store;
```

Comment:
At WebKit layer, we need one object to handle all fetches (to handle pause/resume/click orders coming from UIProcess). The store is the one responsible for that.
We could add intermediary objects for each origin and/or each registration that would interact with that single object, but I am not sure this is worth it.
Also to handle SWServer::clear(SecurityOrigin), we might add a BackgroundFetchStore::clear(SecurityOrigin) since we might lazily load each origin.